src/apps/chifra/pkg/cache/read.go

```diff
@@ -0,0 +1,813 @@
+		return
+	}
+
+	// read items as slice of bytes
+	items := make([]byte, length*8)
+	err = read(&items)
+	if err != nil {
+		return
+	}
+	result := big.NewInt(0)
+	// in order to get the correct number, we have to manually reverse bytes
+	// see https://github.com/golang/go/issues/40891
+	orderedItems := reverseBytes(items)
+	// load bytes into big.Int
+	result.SetBytes(orderedItems)
+	*target = *result
+	return
+}
+
+// makeArrayItemRead takes a function that reads bytes and changes variable's value
+// and turns it into a function returning a value. Such a function can then be provided
+// to `readFromArray` to read and convert array items
+func makeArrayItemRead[Item ArrayItem](
+	reader *bufio.Reader,
+	readAndSetValue func(reader *bufio.Reader, target *Item) (err error),
+) func(reader *bufio.Reader) (value *Item, err error) {
+	return func(reader *bufio.Reader) (value *Item, err error) {
+		v := new(Item)
+		err = readAndSetValue(reader, v)
+		value = v
+		return
+	}
+}
+
+func readTimestamp(read readBytes, target *time.Time) (err error) {
+	var rawTimestamp int64
+	err = read(&rawTimestamp)
+	if err != nil {
+		return
+	}
+	*target = time.Unix(rawTimestamp, 0)
+	return
+}
+
+func readUintAsBool(read readBytes, target *bool) (err error) {
+	var raw uint8
+	err = read(&raw)
+	if err != nil {
+		return
+	}
+	*target = raw != 0
+	return
+}
+
+func readCacheHeader(reader *bufio.Reader, target *cacheHeader) (err error) {
+	read := createReadFn(reader)
+
+	err = read(&target.deleted)
+	if err != nil {
+		return
+	}
+
+	err = read(&target.schema)
+	if err != nil {
+		return
+	}
+
+	err = read(&target.showing)
+	if err != nil {
+		return
+	}
+
+	err = readCString(reader, &target.className)
+	if err != nil {
+		return err
+	}
+
+	return
+}
+
+func ReadBlock(reader *bufio.Reader) (block *types.SimpleBlock, err error) {
+	block = &types.SimpleBlock{}
+	read := createReadFn(reader)
+
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.GasLimit)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.GasUsed)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &block.Hash)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.BlockNumber)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &block.ParentHash)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &block.Miner)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.Difficulty)
+	if err != nil {
+		return
+	}
+
+	err = read(&block.Finalized)
+	if err != nil {
+		return
+	}
+
+	err = readTimestamp(read, &block.Timestamp)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &block.BaseFeePerGas)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &block.Transactions, ReadTransaction)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func ReadTransaction(reader *bufio.Reader) (tx *types.SimpleTransaction, err error) {
```

Comment:
Eventually, we'll move this into the auto-generated type specific file. Please add a TODO, but don't move this yet.